### PR TITLE
[DependencyInjection] Support for intersection types

### PR DESCRIPTION
**Pull request by derrabus**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | #41552
| License       | MIT
| Doc PR        | N/A

### CODE REVIEW

**Comment by derrabus** [comment]
Psalm errors are false positives: Apparently Psalm does not know `ReflectionIntersectionType` yet.

File: src/Symfony/Component/DependencyInjection/Tests/Dumper/PreloaderTest.php [hunk 1]
```diff
@@ -72,4 +75,21 @@ public function testPreloadUnion()
...
         self::assertTrue(class_exists(E::class, false));
     }
+
+    /**
+     * @requires PHP 8.1
+     */
+    public function testPreloadIntersection()
+    {
+        $r = new \ReflectionMethod(Preloader::class, 'doPreload');
+        $r->setAccessible(true);
```
Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Tests/Dumper/PreloaderTest.php]: IIUC this is a no-op in 8.1 :)

Comment (derrabus) [reply src/Symfony/Component/DependencyInjection/Tests/Dumper/PreloaderTest.php]: That is absolutely correct. 🤓

File: src/Symfony/Component/DependencyInjection/LazyProxy/ProxyHelper.php [hunk 2]
```diff
@@ -33,22 +33,31 @@ public static function getTypeHint(\ReflectionFunctionAbstract $r, \ReflectionPa
...
+            $reflectionTypes = $type->getTypes();
+            $glue = '&';
+        } elseif ($type instanceof \ReflectionNamedType) {
+            $reflectionTypes = [$type];
+        } else {
+            return null;
+        }
 
-        foreach ($type instanceof \ReflectionUnionType ? $type->getTypes() : [$type] as $type) {
-            $name = $type instanceof \ReflectionNamedType ? $type->getName() : (string) $type;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/LazyProxy/ProxyHelper.php:38]: on 7.1, don't we need the code path with a cast to string?

Comment (derrabus) [reply src/Symfony/Component/DependencyInjection/LazyProxy/ProxyHelper.php:38]: The string cast was necessary for PHP 7.0 and earlier.

The `if` block above makes sure we have an array of [`ReflectionNamedType`](https://www.php.net/ReflectionNamedType):
* It is the only core implementation of `ReflectionType` on PHP 7.1 - 7.4.
* The `getTypes()` method on `ReflectionUnionType` and `ReflectionIntersectionType` will (as of today) return an array of `ReflectionNamedType`.
* If we encounter another implementation (which a future PHP release might introduce), this method will give up and exit (this is the currently unreachable `else return null;` branch).

It is possible to create userland implementations of `ReflectionType` though, but at this point we should only be dealing with reflection objects coming from PHP directly.

**Comment by nicolas-grekas** [comment]
Thank you @derrabus.

**Pull request merged**